src/probability_theory/probability_space.lean

```diff
@@ -0,0 +1,777 @@
+-/
+
+open measure_theory measurable_space
+open_locale big_operators classical
+
+namespace probability_theory
+
+/- An event is a measurable set in a measurable space that has a probability measure. Note:
+the definition below does not require that the measurable_space have volume one. -/
+@[derive [has_union,has_emptyc,has_compl, has_sdiff,has_union,has_inter,boolean_algebra]]
+def event (Ω : Type*) [measurable_space Ω] : Type* := {x : set Ω // measurable_set x}
+
+/- A probability space is a measure space with volume one. -/
+class probability_space (α : Type*) extends measure_space α := (volume_univ : volume (set.univ) = 1)
+
+export probability_space (volume_univ)
+
+@[simp]
+lemma probability_space.univ_one' {α:Type*} (Pα:probability_space α):
+  (@measure_theory.measure_space.volume α Pα.to_measure_space) (@set.univ α) = 1 :=
+begin
+  rw ← measure_theory.coe_to_outer_measure,
+  rw ← measure_theory.outer_measure.measure_of_eq_coe,
+  simp,
+  rw probability_space.volume_univ,
+end
+
+lemma event_val_eq_coe {Ω : Type*} [probability_space Ω]
+  (X : event Ω) : X.val =
+  (@coe (subtype (@measurable_set Ω _)) (set Ω) _ X) :=
+  by refl
+
+lemma event.eq {Ω : Type*} [probability_space Ω] (A B : event Ω) :
+  A.val = B.val → A = B := subtype.eq
+
+def event_mem {Ω : Type*} [P : probability_space Ω] (a : Ω) (E : event Ω) : Prop :=
+  a ∈ E.val
+
+/- The weight of an event is ≤1. -/
+lemma prob_le_1 {Ω : Type*} [probability_space Ω] (S : set Ω) :
+  volume S ≤ 1 :=
+begin
+  have A3 : volume S ≤ volume set.univ := volume.mono (set.subset_univ _),
+  rw volume_univ at A3,
+  exact A3,
+end
+
+/- The probability of an event is not infinite. -/
+lemma prob_not_infinite {Ω : Type*} [probability_space Ω] (S : set Ω) :
+  (volume S) ≠ ⊤ :=
+begin
+  have A1 : volume S ≤ 1,
+  {  apply prob_le_1,},
+  intro A2,
+  rw A2 at A1,
+  have A3 : (1 : ennreal)=⊤,
+  { apply complete_linear_order.le_antisymm,
+  { apply (ennreal.complete_linear_order.le_top),},
+  { apply A1,}},
+  have A4 : (1 : ennreal) ≠ (⊤ : ennreal),
+  { apply ennreal.one_ne_top,},
+  rw A3 at A4,
+  apply A4,
+  refl,
+end
+
+/- Probabilities are non-negative real numbers. -/
+lemma prob_nnreal {Ω : Type*} [probability_space Ω] (S : set Ω) :
+   ↑((volume S).to_nnreal) = volume S :=
+begin
+  apply ennreal.coe_to_nnreal,
+  apply prob_not_infinite,
+end
+
+/- The probability of an event is equal to its measure. -/
+def event_prob {Ω : Type*} [probability_space Ω] (E : event Ω) : nnreal :=
+  (volume E.val).to_nnreal
+
+/- Pr[E] is the probability of an event. -/
+notation `Pr[`E`]` := event_prob E
+
+lemma event_prob_def {Ω : Type*} [probability_space Ω] (E : event Ω) :
+  ↑(Pr[E]) = (volume E.val) := by
+begin
+  unfold event_prob,
+  apply prob_nnreal,
+end
+
+lemma to_nnreal_almost_monotonic (a b : ennreal) : (a ≠ ⊤) → (b ≠ ⊤) → (a ≤ b)  →
+  (a.to_nnreal ≤ b.to_nnreal) :=
+begin
+  exact λ (ha : a ≠ ⊤) (hb : b ≠ ⊤), (ennreal.to_real_le_to_real ha hb).mpr
+end
```

Comment:
If your proof starts with `intro` or ` λ` usually your colon was too soon:
```suggestion
lemma to_nnreal_almost_monotonic (a b : ennreal) (ha : a ≠ ⊤) (hb : b ≠ ⊤) :
  a ≤ b →a.to_nnreal ≤ b.to_nnreal :=
begin
  exact (ennreal.to_real_le_to_real ha hb).mpr
end
```
At which point you don't need `exact`:

```suggestion
lemma to_nnreal_almost_monotonic (a b : ennreal) (ha : a ≠ ⊤) (hb : b ≠ ⊤) :
  a ≤ b →a.to_nnreal ≤ b.to_nnreal :=
(ennreal.to_real_le_to_real ha hb).mpr
```
At which point you should realize "this is just a worse version of `ennreal.to_real_le_to_real`, there is no point in this lemma"